### PR TITLE
Part 3 of React documentation

### PR DESCRIPTION
**Pull request by BenAtEventbrite**
Has a bunch miscellaneous guidelines based upon our ESLint rules 😄

### CODE REVIEW

File: react/README.md [hunk 1]
```diff
@@ -282,7 +286,7 @@ export default class TextInput extends React.Component {
 
 **[⬆ back to top](#table-of-contents)**
 
-### Ordering
+### `propTypes` Ordering
```
Comment (kaylieEB) [inline react/README.md]: I know it's not part of this commit, but was the extra line on line after `onChange propType` #299 intentional?

File: react/README.md [hunk 2]
```diff
@@ -363,9 +367,75 @@ Coming soon...
...
+    _handleChange(e) {
+        let value = e.target.value;
+
+        if (this.props.onChange) {
+            this.props.onChange(value);
+        }
+    }
+
+    render() {
+        <input type="text" onChange={this._handleChange.bind(this)} />
```
Comment (kaylieEB) [inline react/README.md]: missing `return ()`?

File: react/README.md [hunk 3]
```diff
@@ -568,6 +638,292 @@ When passing a variable to a prop, the curly braces should **not** be padded by
...
```
Comment (kaylieEB) [inline react/README.md]: return?

**Comment by kaylieEB** [comment]
Seems like the 5 of the examples using `<input type="text" onChange={this._handleChange.bind(this)} />` are not in JSX wrapping using `return ()`?

File: react/README.md [hunk 4]
```diff
@@ -568,6 +638,368 @@ When passing a variable to a prop, the curly braces should **not** be padded by
 
 **[⬆ back to top](#table-of-contents)**
 
+## React `key` prop
+
+### Mandatory `key` prop
+
+When rendering an array of React components, specify the `key` prop for each component in the array to aid React's Virtual DOM diffing algorithm (eslint: [`react/jsx-key`](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-key.md)):
```
Comment (vincent-eb) [inline react/README.md]: Maybe add a warning that react will output a console error is this isn't followed? Just to reinforce this is more than just a best practice but important for performance :)

File: react/README.md (hunk 1 above)
Comment (BenAtEventbrite) [inline react/README.md]: Yeah it was visually separating required from optional `propTypes`

**Comment by BenAtEventbrite** [comment]
Thanks @kaylieEB for that catch! Our eslint rules wold've caught it, but there's no eslint in markdown 😉

**Comment by BenAtEventbrite** [comment]
Good call @vincent-eb! Added!

**Pull request merged**